blockchain/utxoio_test.go

```diff
@@ -0,0 +1,548 @@
+// Copyright (c) 2015-2016 The btcsuite developers
+// Copyright (c) 2015-2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package blockchain
+
+import (
+	"bytes"
+	"errors"
+	"os"
+	"path/filepath"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/decred/dcrd/blockchain/stake/v4"
+	"github.com/decred/dcrd/database/v2"
+	"github.com/decred/dcrd/wire"
+)
+
+// TestDbFetchUtxoDatabaseInfo ensures that putting and fetching the utxo
+// database info works as expected.
+func TestDbFetchUtxoDatabaseInfo(t *testing.T) {
+	t.Parallel()
+
+	// Create a test database.
+	dbPath := filepath.Join(os.TempDir(), "test-dbfetchutxodatabaseinfo")
+	_ = os.RemoveAll(dbPath)
+	db, err := database.Create("ffldb", dbPath, wire.MainNet)
+	if err != nil {
+		t.Fatalf("error creating test database: %v", err)
+	}
+	defer os.RemoveAll(dbPath)
+	defer db.Close()
+
+	tests := []struct {
+		name   string
+		dbInfo *utxoDatabaseInfo
+	}{{
+		name:   "without UTXO database info (fresh database)",
+		dbInfo: nil,
+	}, {
+		name: "with UTXO database info",
+		dbInfo: &utxoDatabaseInfo{
+			version: 1,
+			compVer: 2,
+			utxoVer: 3,
+			created: time.Unix(1584246683, 0), // 2020-03-15 04:31:23 UTC
+		},
+	}}
+
+	for _, test := range tests {
+		if test.dbInfo != nil {
+			// Create the UTXO database info bucket.
+			err = db.Update(func(dbTx database.Tx) error {
+				_, err := dbTx.Metadata().CreateBucketIfNotExists(utxoDbInfoBucketName)
+				return err
+			})
+			if err != nil {
+				t.Fatalf("%q: error creating UTXO bucket: %v", test.name, err)
+			}
+
+			// Update the UTXO database info in the database.
+			err = db.Update(func(dbTx database.Tx) error {
+				return dbPutUtxoDatabaseInfo(dbTx, test.dbInfo)
+			})
+			if err != nil {
+				t.Fatalf("%q: error putting UTXO database info: %v", test.name, err)
+			}
+		}
+
+		// Fetch the UTXO database info from the database.
+		err = db.View(func(dbTx database.Tx) error {
+			gotDbInfo, err := dbFetchUtxoDatabaseInfo(dbTx)
+			if err != nil {
+				return err
+			}
+
+			// Ensure that the fetched UTXO database info matches the expected UTXO
+			// database info.
+			if !reflect.DeepEqual(gotDbInfo, test.dbInfo) {
+				t.Errorf("%q: mismatched db info:\nwant: %+v\n got: %+v\n", test.name,
```

Comment:
This probably should be `Fatalf` since all the rest are.

Reply:
I actually did intend this to be different since the other `Fatalf` are for unexpected errors (and we probably shouldn't continue), and this case is just comparing the expected and actual database info.  I don't have a strong opinion, though, and it doesn't matter much since there are only a couple of test cases here, so I just went ahead and updated it.

Reply:
Fair enough.  I wasn't totally convinced one way or the other as I sort of guessed that was maybe the intention, but given the tests depend on order, I thought maybe `Fatalf` was more appropriate and it was like the rest too.  Either way would be fine.  Thanks for updating.